File: src/pass-by-value.js
//See http://jsfiddle.net/7485U/1/

/**
 * The purpose of this file is to display how JS actually passes object references by value
 * rather than passing objects by reference, confirming that JS is actually a pass-by-value language
 * @author Andrew Crites <explosion-pills@aysites.com>
 * @copyright 2012
 */
function barn(foon) {
   var foon = {};
   foon.foon = "barn";
}

var foon = {};
foon.foon = "foon";
barn(foon);

//Should be able to print from command-line JS interpreter without console or alert, browser without console,
//and browsers with console.
if (typeof console == 'undefined') {
   if (typeof alert == 'undefined') {
      print(foon.foon);
   }
   else {
      alert(foon.foon);
   }
}
else {
   console.log(foon.foon);
}

File: src/PassByValue.java
/**
 * The purpose of this file is to display how Java actually passes object references by value
 * rather than passing objects by reference, confirming that Java is actually a pass-by-value language
 * @author Andrew Crites <explosion-pills@aysites.com>
 * @copyright 2012
 */

public class PassByValue {
  public static void main(String [] args) {
    Foon foon = new Foon("foon");
    Foon.barn(foon);
    System.out.println(foon);
  }
}

class Foon {
  private String foon;

  public Foon(String foon) {
    this.foon = foon;
  }

  public static void barn(Foon foon) {
    foon = new Foon("barn");
  }

  public String toString() {
    return foon;
  }
}
